packages/compass-assistant/src/components/tool-call-message.tsx

```diff
@@ -0,0 +1,334 @@
+  padding: spacing[200],
+  borderRadius: '4px',
+  fontFamily: 'Source Code Pro, monospace',
+  whiteSpace: 'pre-wrap',
+  wordBreak: 'break-all',
+});
+
+const codeBlockDarkModeStyles = css({
+  backgroundColor: palette.gray.dark2,
+});
+
+const expandableHeaderStyles = css({
+  display: 'flex',
+  alignItems: 'center',
+  justifyContent: 'space-between',
+  cursor: 'pointer',
+  padding: spacing[200],
+  marginTop: spacing[100],
+  borderRadius: '4px',
+  backgroundColor: palette.gray.light2,
+  transition: 'background-color color 0.2s ease-in-out',
+  '&:hover': {
+    backgroundColor: palette.gray.light1,
+    color: palette.black,
+  },
+});
+
+const expandableHeaderDarkModeStyles = css({
+  backgroundColor: palette.gray.dark2,
+  '&:hover': {
+    backgroundColor: palette.gray.dark1,
+  },
+});
+
+const expandableContentStyles = css({
+  marginTop: spacing[100],
+  padding: spacing[100],
+  borderRadius: '4px',
+  backgroundColor: palette.gray.light2,
+});
+
+const expandableContentDarkModeStyles = css({
+  backgroundColor: palette.gray.dark2,
+});
+
+const buttonGroupStyles = css({
+  display: 'flex',
+  gap: spacing[200],
+  marginTop: spacing[300],
+  '> button': {
+    flex: 1,
+  },
+});
+
+const statusStyles = css({
+  display: 'flex',
+  alignItems: 'center',
+  gap: spacing[100],
+  marginTop: spacing[200],
+});
+
+const statusTextStyles = css({
+  color: palette.gray.dark1,
+  fontWeight: 500,
+  fontSize: '12px',
+});
+
+export interface ToolCallPart {
+  type: string;
+  toolCallId?: string;
+  state?:
+    | 'approval-requested'
+    | 'approval-responded'
+    | 'input-available'
+    | 'output-available'
+    | 'output-error'
+    | 'output-denied';
+  input?: Record<string, unknown>;
+  output?: {
+    content?: Array<{
+      type: string;
+      text?: string;
+    }>;
+  };
+  approval?: {
+    id: string;
+    approved?: boolean;
+    reason?: string;
+  };
+}
+
+interface ToolCallMessageProps {
+  toolCall: ToolCallPart;
+  onApprove?: (approvalId: string) => void;
+  onDeny?: (approvalId: string) => void;
+}
+
+// Extract tool name from type (e.g., "tool-list-databases" -> "list-databases")
+function getToolDisplayName(type: string): string {
+  return type.replace(/^tool-/, '');
+}
+
+export const ToolCallMessage: React.FunctionComponent<ToolCallMessageProps> = ({
+  toolCall,
+  onApprove,
+  onDeny,
+}) => {
+  const darkMode = useDarkMode();
+  const toolsController = useToolsController();
+  const [isInputExpanded, setIsInputExpanded] = useState(false);
+  const [isOutputExpanded, setIsOutputExpanded] = useState(false);
+
+  const toolName = getToolDisplayName(toolCall.type);
+  const description =
+    toolsController.getToolDescription(toolCall.type.replace(/^tool-/, '')) ||
+    '';
+  const inputJSON = JSON.stringify(toolCall.input || {}, null, 2);
+
+  // Extract output text if available
+  const outputText = toolCall.output
+    ? JSON.stringify(toolCall.output, null, 2)
```

Comment:
I'm not sure if we want the context we send to be JSON (so we'd probably format it as EJSON?) or a string. Right now it is formatted as a string with our `toJSString()` helper which seems to work, but that's not exactly standard. And the output shows a bunch of line breaks if you expand it.

Still thinking about it..

Reply:
Makes sense. I'm wondering how well it will grok EJSON as opposed to this pretty-printed "shell script". Because queries aren't written as EJSON, they get written as shell script. I happened to have this example using `null` in my query history which is actually a good one. You can't represent null in standard JSON. In shell script it is just null. And the model seemed to make sense of it.

Like.. what would happen with the nuances of different specific number types?

Probably all edge cases for later, though.